### PR TITLE
Darg pin mods

### PR DESCRIPTION
**Pull request by janm012012**
This PR contains drag pin mods from cherry picked from PR #1 [(vespaman:serial-dma):](https://github.com/vespaman/Smoothieware-CHMT/tree/serial-dma)
- anti sticking wiggle: if the drag pin is not up after timeout, the controller will move around to get it to release
- when releasing the drag pin, the "OK" response is held until the pin is confirmed to be in the up position
- the power management to move the pin down quickly while preventing it from burning.

### CODE REVIEW

**Comment by janm012012** [comment]
Hi @vespaman!

I've cherry picked the drag pin asw and verify/delay on up code form your serial-dma branch and would like you to review it. Is it complete? Does it contain stuff, that's unrelated? What is your recommended machine configuration with this PR merge (still with interlock? What timings do you use for drag-pin down now?) How will uses notice the ASW code (is it hiding badly configured feeders - slowing things down? what happens if ASW fails?)
Thank You!

Jan

**Comment by vespaman** [comment]
Hi Jan,
Hm, it is a bit hard to follow the cherry picking, but..
It looks like you also have my interlock stuff included, did you want that? Regarding ASW, you can read about it in the log, as documented in the MD file. Or did you mean to ask if there's any other way, like a red LED etc?  I remember saying I should put a LED, but honestly, this just works, so I have never bothered. :)

Regarding drag-pin down timings; not sure what you mean? The local drag pin power reduction support or? (That does not seem to be included amongst your cherry's, so did you mean the local up management? Then there's no timing needed anymore, just remove it, controller will manage this altogether. 
You can read about settings in the config.default file, and also, I have included a machine.xml file in the ConfigSamples direcory. (But not sure if I have actually updated it since I fixed the local power management, when I come to think about it).

 - Micael

**Comment by janm012012** [comment]
Hi @vespaman!

Many thanks for your response, that's very much appreciated as all your work on the controller firmware.

I like almost everything you did to the controller firmware. However, I prefer to merge it in chunks that contain a single feature. Recently I learned that by cherry picking into a new branch, a PR can be split into multiple. I started using the drag pin stuff. Unfortunately I did not noticed your power management changes even thought I was informed about the new commit and obviously used an imprecise name for this PR. I was also not able to isolate everything correctly as you at some point renamed the readme which I now cherry picked into this PR as well. Anyhow, I think I shall extend this PR to all your drag pin mods. Looking at the diff between this PR and the head reveals, that the different features seem to be each in its own file. That is probably enough isolation...

Concerning all my questions: as a new user to your mods what changes do you recommend to my machine configuration (I suppose I shall also publish them with the machine.xml in this repository to keep it in sync)

- The ASW probably does not require any changes, or?
- Does the Interlock feature (drag pin release waits until its confirmed to be up) allow us to run without the interlock in OpenPnP?
- The drag pin down G-Code contains a delay to avoid false positive interlocks. I guess this can/shall be removed, or?
- With the drag pin power-management the question will arise what delay you use to let the pin safely go down before you start to move again? (I'll cherry pick that into this PR later)

If you have any other comments please let me know, they are always welcome!

Jan

**Comment by janm012012** [comment]
Hi @vespaman!

I've now cherry picked your drag pin power management and applied to minor modification:

1. activation_start_time == 0 has the special meaning _off_. This is now considered when activating the drag pin.
2. I've changed the calculation of the reduction to use the difference between now and start_time. This is known to work reliable even in case of overflows.

Please review this changes before I merge them into the public version. Thank You!

Jan

**Comment by vespaman** [comment]
Hi @janm012012 

> Hi @vespaman!
> 
> Many thanks for your response, that's very much appreciated as all your work on the controller firmware.
> 
> I prefer to merge it in chunks that contain a single feature. 

Yes, sorry about that! Early on, I did a lot of work in parallel, combined with a) rusty git experience and b) poor memory of checking in. I have since become gradually better though. :-)

> I was also not able to isolate everything correctly as you at some point renamed the readme

Yes, the readme was of the old format, and very outdated. I still don't like it, but at least one can read about the changes against Chris's repo.


> Concerning all my questions: as a new user to your mods what changes do you recommend to my machine configuration (I suppose I shall also publish them with the machine.xml in this repository to keep it in sync)
I think it would be good to have different machine.xml - at least one chmt48 and one chmt36. So are you already saving yours? Did you put it in the ConfigSamples, or anywhere else? Maybe we can find a common place for them.

>     * The ASW probably does not require any changes, or?
No, it should not, if combined with the drag pin up check.

>     * Does the Interlock feature (drag pin release waits until its confirmed to be up) allow us to run without the interlock in OpenPnP?
I think you are perhaps misunderstanding (or me!), but drag pin up management is one thing, and the interlock changes another.
The local drag pin up management does not need any other change in OpenPnP, apart from removing the static delay (wait) that where there previously.

When I talk about interlock changes above, I mean the changes I did to remove any local interlock in the FW. So in my FW, all interlock needs to be done in OpenPnP (which was already the case). 
In my readme, this is mentioned as "Removed machine coordination of actuators", which is the better FW naming to OpenPnP interlock.
In upstream, this meant that in some situations, OpenPnP had to wait for FW for no good reason.

>     * With the drag pin power-management the question will arise what delay you use to let the pin safely go down before you start to move again? (I'll cherry pick that into this PR later)

Yes, I have done some tests, but I don't want to go crazy, as I use my machine for comparatively large batches. Currently it sits at 10ms (which is what it was before, IIRC). I need to find a way to measure the actual time, properly. 

> 
> 
> If you have any other comments please let me know, they are always welcome!

If I understand you correctly, you are basically leaving out the vacuum fix, and the serial? 

The vacuum fix was very simple, and it should make a difference also on the 115.2/460kbit standard bit rates. I would add this also to your tree, while you are at it.

If you have time, I'd also suggest just test the serial-dma, I think that is the most important change, since upstream is so fragile. But maybe one step at a time is a good tactic! 

-  Micael

**Comment by vespaman** [comment]
Hi @janm012012,

> I've now cherry picked your drag pin power management and applied to minor modification:
> 
>     1. activation_start_time == 0 has the special meaning _off_. This is now considered when activating the drag pin.
Good catch!

> 
>     2. I've changed the calculation of the reduction to use the difference between now and start_time. This is known to work reliable even in case of overflows.

OK, if you are sure it will work, then go for it! I'll stick to my version, since in my mind it is easier (and tested). But you did not yet compile that snippet, did you. (There's a spelling error on do_reduction (do_recucsion).


 Micael

**Comment by janm012012** [comment]
Hi @vespaman!

> (There's a spelling error on do_reduction (do_recucsion).

Many thanks for that! I've just fixed it.

> > * Does the Interlock feature (drag pin release waits until its confirmed to be up) allow us to run without the interlock in OpenPnP?
> 
> I think you are perhaps misunderstanding (or me!), but drag pin up management is one thing, and the interlock changes another.
> The local drag pin up management does not need any other change in OpenPnP, apart from removing the static delay (wait) that where there previously.
> 
> When I talk about interlock changes above, I mean the changes I did to remove any local interlock in the FW. So in my FW, all interlock needs to be done in OpenPnP (which was already the case).
> In my readme, this is mentioned as "Removed machine coordination of actuators", which is the better FW naming to OpenPnP interlock.
> In upstream, this meant that in some situations, OpenPnP had to wait for FW for no good reason.

Ok, you mean you actually implemented two things: 1) confirm that the drag pin is up when commanded to do so (I guess the while loop in switch.cpp line 336) and 2) coordinate drag pin going up and down with motion (how/where did you do that?). 

There is this modification in Endstops.cpp lines 1095 - 1115 which I intended to merge but actually don't understand. It is called when "if (gcode->has_m) { switch (gcode->m) { case 119: {" which reads like triggered by M119 to me and I don't noticed yet, that OpenPnP is sending M119. So what does this code do? Is is related to the machine coordination?

> If I understand you correctly, you are basically leaving out the vacuum fix, and the serial?

I'm sorry, that's a misunderstanding. I'd like to merge them in a separate steps. At present I can't open a second PR (easily) because the renaming of Readme makes it difficult to cherry pick that as long as the head is still on the old state. Once this PR is merged, I'll try to cherry pick the vacuum stuff (supposed I find out what they are). Then the diff between head and serial-dma shall be much smaller so that I can understand what you did and why you did it.

Jan

**Comment by vespaman** [comment]
Hi @janm012012,

So the drag pin up works like this; On command request, the actuator is pulled, and then FW waits for the optical sensor to react on the pin up. Only then does the FW send back 'ok' to the OpenPnP, which can then carry on (he's waiting). After this, there's a debounce check going on in the FW, because quite likely the pin will bounce a couple of times, which is OK, but has to be dealt with.

The machine coordination; upstream, the fw waited for motion to stop before activating an actuator, this I removed, since it made the logic up in OpenPnP inconclusive and most of the time wrong. So now, the actuator will be actuated as soon as the command arrive.

Endstop code you are talking about is part of drag pin up management, it is the debounce part, that happens at a later stage. I described it [here](https://groups.google.com/g/openpnp/c/JUCRTa6pKbQ/m/ExoHQ3LoBAAJ)

I have thought about moving my serial dma into my head, but I also thought it (head) might be useful for understanding what drove me into the DMA solution. Maybe it is not useful information any longer, though.

Guess time will tell, what is next in my fork, I have some ideas, but as usual, the time is the problem... :-)
I would guess, that once you have merged my serial-dma with your head, they should be more or less identical, but it will be nice to verify this.

 Micael

**Comment by janm012012** [comment]
Hi @vespaman!

I've reviewed my and your (from the serial-dma branch) machine.xml and learned, that M119 is used to read the drag pin status as part of the axes interlock configured in OpenPnP (to prevent moves while the drag pin is expected to be up and actually isn't).

Now I also understand why you added (and we need) two delays to wait for the drag pin to go up: first when commanded to go up to trigger the ASW if it did not in time and second when OpenPnPs axes interlock is configured to avoid assertions when the pin is bouncing down again.

Concerning the machine coordination, I still don't get it. Yes, in OpenPnP actuators can specify if machine coordination is required. If enabled, a wait (M400 and if Location Confirmation is enabled M114) is inserted. (As we discussed, IMHO this is to be avoided because it breaks the flow). In your machine.xml (in the serial-dma branch) "Before Actuation?" is still enabled for the DRAGPIN actuator and in the code changes I cherry picked, I don't see what you changed/added to force the coordination in the controller. How/where did you implement it?

Jan

**Comment by vespaman** [comment]
Hi @janm012012,

Re: Machine coordintion. No you are reading my explanations wrong! I did not add, I removed. Second bullet point [here](https://github.com/vespaman/Smoothieware-CHMT/tree/serial-dma)

So;
Yes, in OpenPnP, we have the coordination, where we want it. Trouble was in upstream Smoothieware, there's also coordination in the controller. This I removed.
If this is not in your cherry picked code, I apologize - I thought it was. It does hurt to have them still around, just that it will wait for motion to stop before carrying out the actuation.

 - Micael

**Comment by janm012012** [comment]
Hi @vespaman!

Many thanks! Now I got the point and yes, I did see this changes (still in serial_dma) but as they are removed, I was unsure how they would add a feature... I'm with you that it makes sense - and IIRC that's what the NIST standard says - to not coordinate switches with motion implicitly. I guess making this configurable would provide an opportunity to let the controller do the coordination selectively (like for the drag pin and the vacuum valves)... I guess careful validation is needed to make sure the drag pin interlock will still work (maybe this voids the advantage of the controller wise coordination) and the the pick/place dwell times are still respected (IIRC this delays are timed internally in OpenPnP).

One last question that just came to my mind: did you do anything to make sure the ASW is executed immediately when commanded? In other words will this code still work if OpenPnP would operate the dragpin without coordination and send more motion segments while the pin is going up? (With machine coordination done by OpenPnP a M400 will immediately follow and no other move segments will get written into the queue keeping it empty.)

Jan

**Comment by vespaman** [comment]
Hi @janm012012 

I don't really understand your last question. The ASW is only executed whenever the controller receives "drag pin up", and it does not sense it coming up in the predefined time. So there's no command for ASW per se, everything is handled within the controller. An exception handler, if you like.
OpenPnP cannot send anything during this exception handling, since it has not yet received the needed 'ok' from the controller.

Does this answer your question?

 Micael

**Comment by janm012012** [comment]
Hi @vespaman!

After care testing and reviewing I've now merged it. (BTW: while testing I noticed, that feeding speed heavily depends on the jerk setting. With an x jerk of 2e6 (peeler jerk = 0) feeding is kind of slow compared to a machine gun style at 0.)

Many thanks for all your work! I'll cherry pick and merge your other modification next.

Jan
